Allow interpolation to be asynchronous

markscheel opened the pull request:
## Proposed changes

The interpolator receives volume data asynchronously from DgElementArray elements at various time steps.  Before this PR, the interpolator did no work at time `t_0` until it had finished all its work at all times `t < t_0`.  This is correct behavior for AH finding, in which the interpolation points (i.e. the initial guess for the horizon) at time `t_0` are unknown until the horizon is found at all prior times.  However, for many applications (e.g. CCE), the interpolation points are known ahead of time, so interpolation work can be done at multiple time steps simultaneously.

This PR introduces a new type alias `using is_sequential = ` that appears inside the `apply` function of each `InterpolationTarget`.  If `is_sequential::value` is true, then the previous (sequential) behavior
is unchanged. If `is_sequential::value` is false, then interpolations are processed immediately, regardless of the temporal_id, when volume data is received. 

Before this PR, each `InterpolationTarget` stored interpolated data at only one set of interpolation points. Now, each `InterpolationTarget` holds this same data at multiple temporal_ids simultaneously. The commit that makes this change is the largest.

This PR fixes a bug in which sequential interpolation slowed down the interpolator to the point that evolutions would run out of memory, because too much memory was used storing volume data at times that had not yet been interpolated.

### Types of changes:

- [x] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

markscheel committed:
Increase late-call buffer to 1000.

moxcodes commented on src/NumericalAlgorithms/Interpolation/Callbacks/FindApparentHorizon.hpp:
```diff
@@ -88,9 +88,10 @@ struct FindApparentHorizon {
-    if(num_invalid_pts > 0) {
+    const auto& indices_of_invalid_pts =
+        db::get<Tags::IndicesOfInvalidInterpPoints<Metavariables>>(*box);
+    if(indices_of_invalid_pts.find(temporal_id) != indices_of_invalid_pts.end()
```
This line needs to have clang-format re-applied (the `>0` should have the extra space). Optionally the `find` can be a `count`.

moxcodes commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetReceiveVars.hpp:
```diff
@@ -90,42 +92,47 @@ constexpr bool has_fill_invalid_points_with_v =
-        box, [](const gsl::not_null<
-                    db::item_type<Tags::IndicesOfInvalidInterpPoints>*>
+    const gsl::not_null<db::DataBox<DbTags>*> box,
+    const typename Metavariables::temporal_id::type& temporal_id) noexcept {
```
if you can count on `Metavariables::temporal_id` to be a tag, it is better to use `db::item_type` (same below)

moxcodes commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetReceiveVars.hpp:
```diff
@@ -90,42 +92,47 @@ constexpr bool has_fill_invalid_points_with_v =
+          auto& vars_dest = vars_dest_all_times->at(temporal_id);
+          const size_t npts_dest = vars_dest.number_of_grid_points();
+          const size_t nvars = vars_dest.number_of_independent_components;
+          for (auto index : (*indices_of_invalid_points).at(temporal_id)) {
```
does `->` work instead of `(*).`?

kidder commented on src/NumericalAlgorithms/Interpolation/AddTemporalIdsToInterpolationTarget.hpp:
```diff
@@ -47,11 +47,12 @@ struct AddTemporalIdsToInterpolationTarget {
-                            ids,
-                        const db::const_item_type<Tags::CompletedTemporalIds<
-                            Metavariables>>& completed_ids) noexcept {
+        [
```
I think this belongs in the next commit...

markscheel replied:
Yes you are right.

kidder commented on src/NumericalAlgorithms/Interpolation/Callbacks/FindApparentHorizon.hpp:
```diff
@@ -88,9 +88,10 @@ struct FindApparentHorizon {
+    const auto& indices_of_invalid_pts =
+        db::get<Tags::IndicesOfInvalidInterpPoints<Metavariables>>(*box);
+    if(indices_of_invalid_pts.find(temporal_id) != indices_of_invalid_pts.end()
+       and indices_of_invalid_pts.at(temporal_id).size() >0) {
```
formatting?

kidder commented on src/NumericalAlgorithms/Interpolation/Tags.hpp:
```diff
@@ -40,8 +40,10 @@ struct InterpolationTargets {
 namespace Tags {
 
 /// Keeps track of which points have been filled with interpolated data.
+template <typename Metavariables>
```
can we template these tags on the temporal_id::type instead of the Metavariables?

markscheel replied:
Yes, I'm in the process of doing this.

In doing so, I see the following pattern multiple times.
What do you like better:

```
template<typename Metavariables>
void func(const Parallel::ConstGlobalCache<Metavariables>& cache,
          const typename Metavariables::temporal_id::type& temporal_id) noexcept {
      using TemporalId = typename Metavariables::temporal_id::type;
       ... code here that uses Metavariables and TemporalId...
}
```
or
```
template<typename Metavariables, typename TemporalId>
void func(const Parallel::ConstGlobalCache<Metavariables>& cache,
          const TemporalId& temporal_id) noexcept {
   ... code here that uses Metavariables and TemporalId...
}
```

The former is much longer and harder to read, but keeps the explicit relationship between
Metavariables and TemporalId.  I also have versions of the above with a `Requires<...something involving TemporalId...=nullptr>` in the template argument, which makes the former version even longer.

nilsdeppe replied:
looks with option 2), which is what @kidder and I prefer

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetReceiveVars.hpp:
```diff
@@ -249,21 +273,22 @@ struct InterpolationTargetReceiveVars {
-        ](const gsl::not_null<db::item_type<Tags::IndicesOfFilledInterpPoints>*>
+          &temporal_id, &vars_src, &global_offsets
+        ](const gsl::not_null<
+              db::item_type<Tags::IndicesOfFilledInterpPoints<Metavariables>>*>
```
please use the explicit type instead of item_type everywhere

markscheel replied:
Just for my information:  Is the only place where you should use item_type  a place where you don't know the explicit type?

kidder replied:
yes

nilsdeppe replied:
Why not:
```cpp
    [](const gsl::not_null<Variables<vars_tags>*> vars,
       const Strahlkorper<Frame::Inertial>& strahlkorper,
       const FastFlow& fast_flow) noexcept
```
this is shorter, and less work from the compiler

markscheel replied:
Ah, that is in FindApparentHorizon.hpp. Missed it before. Now made that change too.

kidder commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetReceiveVars.hpp:
```diff
@@ -278,27 +303,42 @@ struct InterpolationTargetReceiveVars {
+        db::get<Tags::IndicesOfFilledInterpPoints<Metavariables>>(box)
+            .at(temporal_id)
+            .size();
+    const size_t invalid_size = [&box,&temporal_id]() noexcept {
```
formatting?

moxcodes commented on src/NumericalAlgorithms/Interpolation/InterpolationTargetReceiveVars.hpp:
```diff
@@ -278,27 +303,42 @@ struct InterpolationTargetReceiveVars {
+    const size_t invalid_size = [&box,&temporal_id]() noexcept {
+      const auto& invalid_indices =
+          db::get<Tags::IndicesOfInvalidInterpPoints<Metavariables>>(box);
+      if (invalid_indices.find(temporal_id) != invalid_indices.end()) {
```
could use `count` for the existence test

kidder commented on src/NumericalAlgorithms/Interpolation/AddTemporalIdsToInterpolationTarget.hpp:
```diff
@@ -42,9 +42,14 @@ struct AddTemporalIdsToInterpolationTarget {
         db::get<Tags::TemporalIds<Metavariables>>(box).empty();
 
+    // Some of the temporal_ids may not be new;
+    // i.e. AddTemporalIdsToInterpolationTarget may have already be
```
be -> been

moxcodes commented on src/NumericalAlgorithms/Interpolation/Tags.hpp:
```diff
@@ -40,8 +40,10 @@ struct InterpolationTargets {
+template <typename Metavariables>
 struct IndicesOfFilledInterpPoints : db::SimpleTag {
-  using type = std::unordered_set<size_t>;
+  using type = std::unordered_map<typename Metavariables::temporal_id::type,
```
`db::item_type` would probably be better (and below)

kidder replied:
nooooooooooooooo

moxcodes replied:
edit: oh, I see-- in your review you suggested templating on the temporal id type. Yes, I agree that is better.

moxcodes commented on tests/Unit/NumericalAlgorithms/Interpolation/Test_AddTemporalIdsToInterpolationTarget.cpp:
```diff
@@ -128,43 +130,82 @@ SPECTRE_TEST_CASE("Unit.NumericalAlgorithms.InterpolationTarget.AddTemporalIds",
+    // For non-sequential, there should be two queued_simple_actions,
+    // each of which will call MockComputeTargetPoints for one of the
+    // new temporal_ids.
+    for(size_t i=0;i<2;++i) {
```
the `for` formatting makes me think this should also get another clang-format.

moxcodes reviewed (commented):
mostly nitpicks from me

markscheel committed:
InterpolationTarget: indices/vars are per temporal_id.

Previously all interpolation was sequential, so we needed
only one set of indices and one set of vars.  But if interpolation
is asynchronous there may be interpolations for multiple
temporal_ids happening at the same time.

markscheel committed:
Can now do asynchronous interpolation.

markscheel commented:
Added a bunch of fixup commits (and reordered them for easier eventual squash).

kidder commented on src/NumericalAlgorithms/Interpolation/AddTemporalIdsToInterpolationTarget.hpp:
```diff
@@ -23,35 +23,31 @@ namespace Actions {
+            typename ArrayIndex, typename TemporalId,
             Requires<tmpl::list_contains_v<
-                DbTags, typename Tags::TemporalIds<Metavariables>>> = nullptr>
+                DbTags, typename Tags::TemporalIds<TemporalId>>> = nullptr>
```
is this typename needed?

markscheel replied:
Nope.  Removing it...

kidder reviewed (commented):
You can squash once @moxcodes is happy and rebase on develop

moxcodes reviewed (commented):
I'm happy for you to go ahead and squash.

markscheel commented:
Rebased, squashed, and removed more db::item_types and removed an unnecessary 'typename' keyword.

nilsdeppe reviewed (approved):
![image](https://user-images.githubusercontent.com/13531030/76662022-40148900-6553-11ea-843a-2c44a8b6b322.png)

nilsdeppe commented:
@moxcodes please let us know if you're happy with this :)
